### ISSUE TITLE
Suggest replacing `result.map_or_else(Err, f)` with `result.and_then(f)`

### ISSUE THREAD

**Issue by ada4a** [rust-lang/rust-clippy#17322]
### Description

This could be a new case in `unnecessary_result_map_or_else`, or a new lint.

### Advantages
- more concise
- easier to understand (it legit took me a couple minutes to decipher the thing) 

### Example
```rust
fn foo(res: Result<Vec<i32>, String>) -> Result<i32, String> {
    res.map_or_else(Err, |mut v| v.pop().ok_or_else(|| "empty list!".to_string()))
}
```

Could be written as:

```rust
fn foo(res: Result<Vec<i32>, String>) -> Result<i32, String> {
    res.and_then(|mut v| v.pop().ok_or_else(|| "empty list!".to_string()))
}
```

### Version

```text
rustc 1.96.0 (ac68faa20 2026-05-25)
binary: rustc
commit-hash: ac68faa20c58cbccd01ee7208bf3b6e93a7d7f96
commit-date: 2026-05-25
host: x86_64-unknown-linux-gnu
release: 1.96.0
LLVM version: 22.1.2
```

### Additional Labels

@rustbot label C-feature-request

**Comment by hiSandog** [comment]
This lint looks useful, but the applicability should be conservative around evaluation order and type inference. A good test set would include closures with side effects, `Err` branches that allocate or borrow, and cases where `map_or_else(Err, f)` helps inference while `and_then(f)` does not. If all of those stay machine-applicable, the suggestion can be much more confident.

**Comment by ada4a** [comment]
> `Err` branches that allocate or borrow

I don't think this one can be an issue, given that the `Err` branch would do nothing but wrap the provided expression in an `Err`.

Could be that I'm misunderstanding you though. Do you have an example in mind of how this could happen? 

> cases where `map_or_else(Err, f)` helps inference while `and_then(f)` does not

Again, I'm struggling to imagine such a case. Would you have an example?

**Comment by mi2ebi** [comment]
the Err branch could be a {} block that does arbitrary things and then returns an Err

**Comment by ada4a** [comment]
I wouldn't lint those cases at all – just the ones where the Err branch is a literal `Err`
